src/mapping/query/functions.rs

```diff
@@ -1,11 +1,98 @@
-use super::Function;
+#[cfg(test)]
+use super::Literal;
+use super::{ArgumentList, Function, Parameter};
 use crate::{
-    event::{Event, Value},
+    event::{Event, Value, ValueKind},
     mapping::Result,
     types::Conversion,
 };
 use bytes::Bytes;
 use chrono::{TimeZone, Utc};
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+// If this macro triggers, it means the logic to detect invalid types did not
+// function as expected. This is a bug in the implementation.
+macro_rules! unexpected_type {
+    ($value:expr) => {
+        unimplemented!("unexpected value type: '{}'", $value.to_value_kind());
```

Comment:
Should this be `unreachable!` instead? You've verified the parameter types previously.